File: sorting/LC238_product_of_array_except_self/solution.ts
function productExceptSelf(nums: number[]) {
  const output = [1];
  output[0] = 1;
  for (let i = 1; i < nums.length; i++) {
      output.push(output[i-1] * nums[i-1])
  }
  let right = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
      output[i] *= right;
      right *= nums[i];
  }
  return output;
}

function LC238Main() {
  const nums = [1,2,3,4]

  const output = productExceptSelf(nums);
  console.log(output);
}

LC238Main()

File: sorting/LC238_product_of_array_except_self/solution.py
class Solution(object):
  def productExceptSelf(self, nums):
    output = [1]
    for i in range(1, len(nums)):
      output.append(output[i-1] * nums[i-1])
    right = 1
    for i in range(len(nums)-1, -1, -1):
      output[i] *= right
      right *= nums[i]
    return output
  
def main():
  nums = [1,2,3,4]
  s = Solution()
  output = s.productExceptSelf(nums)
  print(output)

main()